### ISSUE TITLE
Gradient Calculation

### ISSUE THREAD

**Issue by stellarpower** [AISViz/Soft-DTW#6]
Hey,

I'm trying to implement the backward pass explicitly, using the equation from the paper (and other repos' implementations), in an effort to improve the speed.

If I understand correctly, as the master branch here doesn't include a custom gradient, tensorflow will be using its automatic differentiator to compute the gradients.

However, obviously the algorithm for the forward pass is quite complicated - we have loops, and the softmin is implemented in Cython, which wouldn't be automatically differentiable (although maybe this has no effect on the gradient). I'm therefore wondering, do we know if the gradients tensorflow computes automatically are correct? Have they been verified thus far and checked to be numerically close to computing using the explicit expression? 

Or am I missing something, and it's calculated a different way?

Thanks

**Comment by stellarpower** [comment]
So, I have been in the process of implementing the backwards pass calculations in a branch [here](  https://github.com/stellarpower/Soft-DTW/tree/stellapower.AddFeature.BetterParallelisation  ), and as I have been testing that, I've noticed that for the gradients, there’s a discrepancy between the implementation here and for at least one of the Torch implementations (in this case, [this version, using a numba CUDA kernel](   https://github.com/toinsson/pysdtw/issues  )

The losses I get for each sequence in a batch are identical (first row), but the gradients are all zeros bar the last (second row).

| My branch | master | Torch Implementation |
| -------- | -------- |-------- |
|  | ![image](https://github.com/AISViz/Soft-DTW/assets/5004545/d22ad0a8-efba-43d2-9cb0-3a1b5f79c153) |  ![image](https://github.com/AISViz/Soft-DTW/assets/5004545/fa20216f-0064-48ef-a33c-ef681a9b1d54)   |
| ![image](https://github.com/AISViz/Soft-DTW/assets/5004545/cfce328f-3c8e-4459-bb12-9226bf6ae270)  | ![image](https://github.com/AISViz/Soft-DTW/assets/5004545/42042526-00bf-4cd9-abb5-5d803acee0c6) | ![image](https://github.com/AISViz/Soft-DTW/assets/5004545/0b3f6db1-1e63-44a2-838e-fa1bedbbca72) |

Given that there are so many zeros - and that amazingly it seems my implementation produced the same numbers - I suspect it may be the expression the auto-differentiator comes up with is not valid, which I think would then invalidate use of the loss function in general - unless something else is going on here, e.g. loss reductions(?)

**Comment by gabrielspadon** [comment]
Hi @stellarpower, thanks for the feedback.
The implementation follows https://arxiv.org/abs/1703.01541, https://rtavenar.github.io/ml4ts_ensai/contents/align/softdtw.html, and https://github.com/mblondel/soft-dtw. We will review the results next week to check for anything we let pass. Can you share the full pipeline for testing you created? It can be through an external repository we can clone and test if you prefer.

**Comment by stellarpower** [comment]
Okay, great, thanks. My testing setup was not exactly rigorous at this stage, but [here's](  https://gist.github.com/stellarpower/54c54a85b1fc57c3bccaaa70361d179f ) a bit of a messy test setup. Hopefully it's reasonable to follow, but let me know if anything needs explanation. The testing file for Soft-DTW is [in my branch]( https://github.com/stellarpower/Soft-DTW/blob/stellapower.AddFeature.BetterParallelisation/TestImplementation.py ).

Currently I'm trying to see if the performance can be improved (conversation #4), as TF does not seem to be parallelising that well. Still debating the idea of just writing the kernel in SYCL and a custom TF op.
